### ISSUE TITLE
Gentoo build failure: posixmodule.c:(.text+0x7d9): undefined reference to `makedev'

### ISSUE THREAD

**Issue by flussence** [oilshell/oil#67]
I'm attempting to package/build osh 0.3.0 on Gentoo (I didn't see anyone else trying…) and I get that error. Looks like you need to `#include <sys/sysmacros.h>` somewhere.

Text leading up to failure - there's a ton of warnings above it but they look unrelated:
```
Modules/readline.c: In function ‘flex_complete’:
Modules/readline.c:898:31: warning: passing argument 1 of ‘completion_matches’ discards ‘const’ qualifier from pointer target type [-Wdiscarded-qualifiers]
     return completion_matches(text, *on_completion);
                               ^~~~
Modules/readline.c:39:15: note: expected ‘char *’ but argument is of type ‘const char *’
 extern char **completion_matches(char *, rl_compentry_func_t *);
               ^~~~~~~~~~~~~~~~~~
/var/tmp/portage/app-shells/oil-0.3.0/temp/ccynXJSD.o: In function `posix_makedev':
posixmodule.c:(.text+0x7d9): undefined reference to `makedev'
/var/tmp/portage/app-shells/oil-0.3.0/temp/ccynXJSD.o: In function `posix_minor':
posixmodule.c:(.text+0x858): undefined reference to `minor'
/var/tmp/portage/app-shells/oil-0.3.0/temp/ccynXJSD.o: In function `posix_major':
posixmodule.c:(.text+0x9d8): undefined reference to `major'
collect2: error: ld returned 1 exit status

real    1m0.086s
user    1m3.917s
sys     0m3.014s
/var/tmp/portage/app-shells/oil-0.3.0/work/oil-0.3.0
cat _build/oil/ovm _build/oil/bytecode.zip > _bin/oil.ovm
cat: _build/oil/ovm: No such file or directory
make: *** [Makefile:281: _bin/oil.ovm] Error 1
make: *** Deleting file '_bin/oil.ovm'
```

**Comment by andychu** [comment]
OK great thanks.  I think I can remove those functions from `posixmodule.c`.

As you may have noticed, Oil includes all of Python's `posixmodule.c`, but it doesn't need all of it.  Apparently some macros like `major()` live in different headers on different platforms (this is news to me).

But since Oil doesn't use those macros, I can remove them and simplify the build.

**Comment by andychu** [comment]
Hm actually I was able to build it.  Or actually I built 0.4.alpha1.  I doubt I changed anything that fixed Gentoo in this release.  Can you try 0.4.alpha1 on gentoo anyway?

http://www.oilshell.org/releases.html

I got some related warnings, but it still built.

The errors are `posixmodule.c:(.text+0x7d9): undefined reference to `makedev'`.  Do you need libc development packages?  I'm not sure how Gentoo works.


```
lisa /src/oil-0.4.alpha1 # make
build/compile.sh build-opt _build/oil/ovm _build/oil/module_init.c _build/oil/main_name.c _build/oil/c-module-srcs.txt
/src/oil-0.4.alpha1/Python-2.7.13 /src/oil-0.4.alpha1
Modules/posixmodule.c: In function ‘posix_major’:
Modules/posixmodule.c:7077:13: warning: In the GNU C Library, "major" is defined
 by <sys/sysmacros.h>. For historical compatibility, it is
 currently defined by <sys/types.h> as well, but we plan to
 remove this soon. To use "major", include <sys/sysmacros.h>
 directly. If you did not intend to use a system-defined macro
 "major", you should undefine it after including <sys/types.h>.
     return PyInt_FromLong((long)major(device));
             ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~                                                                                                                                                                                                                                                                                                                                     
Modules/posixmodule.c: In function ‘posix_minor’:
Modules/posixmodule.c:7090:13: warning: In the GNU C Library, "minor" is defined
 by <sys/sysmacros.h>. For historical compatibility, it is
 currently defined by <sys/types.h> as well, but we plan to
 remove this soon. To use "minor", include <sys/sysmacros.h>
 directly. If you did not intend to use a system-defined macro
 "minor", you should undefine it after including <sys/types.h>.
     return PyInt_FromLong((long)minor(device));
             ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~                                                                                                                                                                                                                                                                                                                                     
Modules/posixmodule.c: In function ‘posix_makedev’:
Modules/posixmodule.c:7103:13: warning: In the GNU C Library, "makedev" is defined
 by <sys/sysmacros.h>. For historical compatibility, it is
 currently defined by <sys/types.h> as well, but we plan to
 remove this soon. To use "makedev", include <sys/sysmacros.h>
 directly. If you did not intend to use a system-defined macro
 "makedev", you should undefine it after including <sys/types.h>.
     return _PyInt_FromDev(makedev(major, minor));
             ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~                                                                                                                                                                                                                                                                                                                                         
Modules/signalmodule.c: In function ‘trip_signal’:
Modules/signalmodule.c:180:9: warning: ignoring return value of ‘write’, declared with attribute warn_unused_result [-Wunused-result]
         write(wakeup_fd, "\0", 1);
         ^~~~~~~~~~~~~~~~~~~~~~~~~
Modules/readline.c: In function ‘flex_complete’:
Modules/readline.c:898:31: warning: passing argument 1 of ‘completion_matches’ discards ‘const’ qualifier from pointer target type [-Wdiscarded-qualifiers]
     return completion_matches(text, *on_completion);
                               ^~~~
Modules/readline.c:39:15: note: expected ‘char *’ but argument is of type ‘const char *’
 extern char **completion_matches(char *, rl_compentry_func_t *);
               ^~~~~~~~~~~~~~~~~~

real    0m31.796s
user    0m30.437s
sys     0m1.342s
/src/oil-0.4.alpha1
cat _build/oil/ovm _build/oil/bytecode.zip > _bin/oil.ovm
chmod +x _bin/oil.ovm
```

**Comment by flussence** [comment]
Same error on 0.4.alpha1 and 0.4.0, unfortunately.

Gentoo doesn't have separate dev packages like Debian etc. — everything installs headers by default. Usually it's supposed to be as close to vanilla upstream as possible but I wouldn't be surprised if both their glibc and python2 packages were being different just for the sake of it…

Dug a bit and found [this](https://bugs.gentoo.org/575232) which confirms it. Seems like one developer has a few screws loose and everyone else gets to pay for it.

**Comment by andychu** [comment]
Hm but why was I able to build it inside the Gentoo chroot then?  I just did `emerge --sync` and `./configure; make` and it worked.  I used this to set up the chroot:

https://github.com/oilshell/oil/blob/master/test/gentoo.sh

**Comment by flussence** [comment]
The last update in that bug seems to be an attempt to partially restore sanity, but it's only been applied to the stable version. The one you have there is likely that version, while I'm running unstable, so I guess that explains it.

If nothing else, at least I have a few ideas for workarounds now. Using musl headers is an option...

**Comment by andychu** [comment]
OK that is unfortunate.  I think I will eventually remove the makedev stuff anyway, since OSH doesn't even use it.  But that will come as a big overhaul of `posixmodule.c`, possibly removing it altogether.

At some point I would like to provide binaries statically linked against musl, which should work on any distro easily.  Everything the shell uses is portable and old so this shouldn't be a problem.

Closing for now, but if you have issues on another system let me know.

The rootfs I downloaded is labeled 20180116, which is a lot more recent than that bug.
